File: nihongo-it-backend/bin/main/com/example/nihongoit/service/ConversationService.kt
package com.example.nihongoit.service

import com.example.nihongoit.dto.ConversationDTO
import com.example.nihongoit.dto.CreateConversationRequest
import com.example.nihongoit.dto.PagedResponse
import com.example.nihongoit.dto.UpdateConversationRequest
import org.springframework.data.domain.Pageable
import java.util.UUID

interface ConversationService {
    /**
     * Get all conversations with pagination
     */
    fun getAllConversations(pageable: Pageable): PagedResponse<ConversationDTO>
    
    /**
     * Search conversations by title or description
     */
    fun searchConversations(query: String, pageable: Pageable): PagedResponse<ConversationDTO>
    
    /**
     * Get conversations by JLPT level
     */
    fun getConversationsByJlptLevel(level: String, pageable: Pageable): PagedResponse<ConversationDTO>
    
    /**
     * Get conversation by ID
     */
    fun getConversationById(conversationId: UUID): ConversationDTO
    
    /**
     * Create a new conversation
     */
    fun createConversation(request: CreateConversationRequest): ConversationDTO
    
    /**
     * Update an existing conversation
     */
    fun updateConversation(conversationId: UUID, request: UpdateConversationRequest): ConversationDTO
    
    /**
     * Delete a conversation
     */
    fun deleteConversation(conversationId: UUID)
} 

File: frontend-user/src/services/conversation.service.ts
import api from '@/lib/api'
import type { Conversation } from '@/types/conversation.types'
import type { PagedResponse } from '@/types/common.types'

const conversationService = {
  getConversations(
    page = 0,
    size = 12,
    search = '',
    sortBy = 'title',
    sortDir: 'asc' | 'desc' = 'asc',
  ): Promise<PagedResponse<Conversation>> {
    const params: Record<string, string | number> = { page, size, sortBy, sortDir }
    if (search.trim()) params.search = search
    return api.get('/api/v1/learning/conversations', { params }).then((r) => r.data)
  },

  getConversationById(id: string): Promise<Conversation> {
    return api.get(`/api/v1/learning/conversations/${id}`).then((r) => r.data)
  },

  getConversationsByJlptLevel(
    level: string,
    page = 0,
    size = 12,
  ): Promise<PagedResponse<Conversation>> {
    return api
      .get(`/api/v1/learning/conversations/jlpt/${level}`, { params: { page, size } })
      .then((r) => r.data)
  },

}

export default conversationService
